determined-ai/determined
# fix: agent config precedence

## Description
This PR makes determined-agent initialization and start-up more similar to determined-master by incorporating viper into agent configuration setup to ensure the correct priority of agent configuration options (flag > environment > config > default).

**It is also worth noting that the the `determined-agent` flag `enable_api` was changed to `api_enabled`.**

## Test Plan
- First, `cd` into local `determined` repo and spin up a cluster with the `devcluster` tool. In the cluster logs in your terminal tab, ensure that the `agent configuration` outputs shows `"agent_reconnect_attempts":5` (provided that the `agent_reconnect_attempts` field is not defined in the `agent.config` section of your devcluster YAML, the JSON output should show the default value, which is currently 5).
- Now, add `agent_reconnect_attempts: 9` to your devcluster config as follows: 
```
config_file:
   ...
    agent: 
         ...
         agent_reconnect_attempts: 9
```
- Spin up your devcluster using this updated config, and verify that the `agent configuration` outputs shows `"agent_reconnect_attempts":9` in your logs.
- Now, press `2` in your terminal to kill the agent binary so that you are only running `DB` and `MASTER`
- Open up a new terminal tab and run `export DET_AGENT_RECONNECT_ATTEMPTS=10`
- Then, `cd` into `determined/agent/build` and run `./determined-agent --config-file /private/tmp/devcluster/agent.conf`
- verify that the `agent configuration` outputs shows `"agent_reconnect_attempts":10` (rather than the current default value of 5 or config value of 9)
- Now, to ensure that flags take precedent over defaults and environment variables, run `./determined-agent --config-file /private/tmp/devcluster/agent.conf --agent-reconnect-attempts 12`
- ensure that the `agent configuration` outputs shows `"agent_reconnect_attempts":12` rather than the default value 5, config value 9, or the environment variable value 10.


## Checklist

  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.

## Ticket
DET-9800
